### PR TITLE
Wrapper class for samples

### PR DESCRIPTION
**Pull request by aagrande**
This adds a simple wrapper class for the main output of Treeffuser: the samples. 

## Subclassing and custom methods

**EDIT**: see update below by @ANazaret .
~~`Samples` subclasses `ndarray`.~~ `Samples`:
- ~~uses view casting to cast an existing standard `ndarray` as an instance of `Samples`~~; 
- ~~ensures downstream compatibility with the NumPy API by preserving the same function signatures of existing methods (details [here](https://numpy.org/doc/stable/user/basics.subclassing.html#subclassing-and-downstream-compatibility))~~; 
- adds methods to simplify the computation of sample statistics (e.g. `sample_mean`) and custom methods to compute correlation matrices, confidence intervals, range, etc..

~~See NumPy's documentation on subclassing for further [reference](https://numpy.org/doc/stable/user/basics.subclassing.html).~~

## Usage
The goal is to facilitate the user experience when computing estimates from Treeffuser samples, such as mean, quantiles, correlation matrices, modes, etc.

```python
samples = model.sample(x, n_samples=100)

samples = Samples(samples)
samples.sample_mean() # conditional mean
samples.sample_quantile(q=[0.05, 0.95])  # conditional quantiles
samples.sample_kde(bandwidth="scott) # conditional KDEs
...
```

### CODE REVIEW

File: src/treeffuser/_base_tabular_diffusion.py [hunk 1]
```diff
@@ -432,7 +432,7 @@ def predict_distribution(
 
         y_samples = self._sample_without_validation(X=X, n_samples=n_samples, verbose=verbose)
 
-        n_samples, batch, _ = y_samples.shape
+        batch = y_samples.shape[2]
```
Comment (ANazaret) [inline src/treeffuser/_base_tabular_diffusion.py:435]: @aagrande It looks like it should be `shape[1]`

**Review by ANazaret — CHANGES_REQUESTED** [review]

File: src/treeffuser/samples.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+def _check_unidimensional(array) -> None:
+    if array.ndim > 2 and array.shape[-1] > 1:
+        raise ValueError("This method only applies to unidimensional responses.")
+
+
+###################################################
+# Main class
+###################################################
+class Samples:
+    def __init__(self, input_array: Float[np.ndarray, "n_samples batch y_dim"]):
```
Comment (ANazaret) [inline src/treeffuser/samples.py:23]: Do you want to add documentation to the class? to explain what's the point of it

File: src/treeffuser/_base_tabular_diffusion.py (hunk 1 above)
Comment (aagrande) [reply src/treeffuser/_base_tabular_diffusion.py:435]: good catch

**Review by ANazaret — APPROVED** [review]

**Comment by ANazaret** [comment]
Thanks @aagrande ! For the record, we chose to not define Samples as a subclass of np.array because the numpy inherited functions were enabling the user to transform a Samples object into a corrupted state: example taking the transpose was returning a non-sensical Samples object.

**Pull request merged**